### PR TITLE
Issue 2127 remove react function redundancies

### LINKED ISSUE #2127: Remove redundancies in state reaction code
Currently the reaction code for states include code for all transitions and local reactions along the state hierarchy. This especially leads to redundancy if statecharts are deeply nested. 

To avoid redundancies the generated code for reaction handling must be refactored in a way that introduces functions for reaction handling also for composite states that can be called from the child state handlers. Additionally it must be possible to inline these functions if no redundancy will be introduced. This would avoid unnecessary functions. 

DoD
- [ ] Refactor ModelSequencer to create required functions
- [ ] Extend Interpreter
- [ ] Extend C code generator (own task)
- [ ] Extend C++ code generator (own task)
- [ ] Extend Java code generator (own task)
- [ ] Extend Python code generator (own task)
- [ ] Extend TypeScript code generator (own task)
- [ ] Extend C# code generator (own task)

### PR DESCRIPTION
**Pull request by terfloth**
fixes #2127

### CODE REVIEW

**Comment by BeckmaR** [comment]
Is it possible to split this up somehow?

**Comment by terfloth** [comment]
retest this please

File: plugins/org.yakindu.sct.model.sexec/src/org/yakindu/sct/model/sexec/transformation/ng/ReactMethod.xtend [hunk 1]
```diff
@@ -0,0 +1,368 @@
...
+	
+	
+	def defineReactMethod(ExecutionState execState) {
+		
+		val state = execState.sourceElement as RegularState
+		val reactMethod = execState.reactMethod
+		val childFirst = state.statechart.isChildFirstExecution
+				
+		val shouldProcessParent = 
+			if (! childFirst)
```
Comment (BeckmaR) [inline plugins/org.yakindu.sct.model.sexec/src/org/yakindu/sct/model/sexec/transformation/ng/ReactMethod.xtend]: I'm not sure how this works. Where does `sv` actually come from? Isn't it null?

File: plugins/org.yakindu.sct.generator.c/src/org/yakindu/sct/generator/c/submodules/InternalFunctionsGenerator.xtend [hunk 2]
```diff
@@ -82,10 +84,37 @@ class InternalFunctionsGenerator {
...
 		«enterSequenceFunctions.toPrototypes»
 		«exitSequenceFunctions.toPrototypes»
-		«reactFunctions.toPrototypes»
+		«reactFunctions.filter[ f | ! (f.eContainer instanceof ExecutionState)].toList.toPrototypes»
+		«reactMethods.toDeclarations»
 		static void «clearInEventsFctID»(«scHandleDecl»);
 		static void «clearOutEventsFctID»(«scHandleDecl»);
 	'''
+
+//	def functionPrototypes(ExecutionFlow it) '''
```
Comment (BeckmaR) [inline plugins/org.yakindu.sct.generator.c/src/org/yakindu/sct/generator/c/submodules/InternalFunctionsGenerator.xtend]: Why is this commented and where does it come from?

File: plugins/org.yakindu.sct.generator.c/src/org/yakindu/sct/generator/c/submodules/InternalFunctionsGenerator.xtend [hunk 3]
```diff
@@ -110,9 +139,37 @@ class InternalFunctionsGenerator {
 		«exitActionFunctions.toImplementation»
 		«enterSequenceFunctions.toImplementation»
 		«exitSequenceFunctions.toImplementation»
-		«reactFunctions.toImplementation»
+		«reactFunctions.filter[ f | ! (f.eContainer instanceof ExecutionState)].toList.toImplementation»
+		«reactMethods.toDefinitions»
 		
 	'''
+
+//	def functionImplementations(ExecutionFlow it) '''
```
Comment (BeckmaR) [inline plugins/org.yakindu.sct.generator.c/src/org/yakindu/sct/generator/c/submodules/InternalFunctionsGenerator.xtend]: Same, why comment?

File: test-plugins/org.yakindu.sct.generator.c.test/all C generator tests.launch [hunk 4]
```diff
@@ -15,13 +15,13 @@
...
 <listAttribute key="org.eclipse.debug.core.MAPPED_RESOURCE_PATHS">
-<listEntry value="/org.yakindu.sct.generator.c.test/test-gen/org/yakindu/sct/generator/c/test/AllTests.java"/>
+<listEntry value="/org.yakindu.sct.generator.c.test/src/org/yakindu/sct/generator/c/test/AllCTests.java"/>
 </listAttribute>
 <listAttribute key="org.eclipse.debug.core.MAPPED_RESOURCE_TYPES">
 <listEntry value="1"/>
 </listAttribute>
 <mapAttribute key="org.eclipse.debug.core.environmentVariables">
-<mapEntry key="GTEST_DIR" value="/usr/local/Cellar/gtest"/>
+<mapEntry key="GTEST_DIR" value="/Users/terfloth/Tools/gtest-1.7.0-bin"/>
```
Comment (tkutz) [inline test-plugins/org.yakindu.sct.generator.c.test/all C generator tests.launch]: ;)

File: plugins/org.yakindu.sct.model.sexec/src/org/yakindu/sct/model/sexec/transformation/ng/ReactMethod.xtend (hunk 1 above)
Comment (BeckmaR) [reply plugins/org.yakindu.sct.model.sexec/src/org/yakindu/sct/model/sexec/transformation/ng/ReactMethod.xtend]: Got it, fancy! But maybe it's a tad too fancy to use it only once.

File: plugins/org.yakindu.sct.generator.c/src/org/yakindu/sct/generator/c/submodules/InternalFunctionsGenerator.xtend (hunk 2 above)
[the remark this answers is not in the document]
Comment (terfloth) [reply plugins/org.yakindu.sct.generator.c/src/org/yakindu/sct/generator/c/submodules/InternalFunctionsGenerator.xtend]: was old implementation - is removed

File: plugins/org.yakindu.sct.generator.c/src/org/yakindu/sct/generator/c/submodules/InternalFunctionsGenerator.xtend (hunk 3 above)
[the remark this answers is not in the document]
Comment (terfloth) [reply plugins/org.yakindu.sct.generator.c/src/org/yakindu/sct/generator/c/submodules/InternalFunctionsGenerator.xtend]: was old implementation - is removed

File: test-plugins/org.yakindu.sct.generator.c.test/all C generator tests.launch (hunk 4 above)
Comment (terfloth) [reply test-plugins/org.yakindu.sct.generator.c.test/all C generator tests.launch]: recovered old path value

File: plugins/org.yakindu.sct.model.sexec/src/org/yakindu/sct/model/sexec/transformation/ng/ReactMethod.xtend (hunk 1 above)
Comment (terfloth) [reply plugins/org.yakindu.sct.model.sexec/src/org/yakindu/sct/model/sexec/transformation/ng/ReactMethod.xtend]: Reviews should be fun & educational ;-)

Comment (terfloth) [reply plugins/org.yakindu.sct.model.sexec/src/org/yakindu/sct/model/sexec/transformation/ng/ReactMethod.xtend]: The construct was copied from the original sequence construction (ReactionBuilder) where it is used at multiple times. 

Inlined it into the processParent expression.

**Comment by andreasmuelder** [comment]
@terfloth please resolve conflicts

File: plugins/org.yakindu.sct.model.sexec/src/org/yakindu/sct/model/sexec/naming/DefaultNamingService.xtend [hunk 5]
```diff
@@ -168,6 +179,7 @@ class DefaultNamingService implements INamingService {
 		var HashMap<NamedElement, String> map = new HashMap<NamedElement, String>
 		map.addShortStateNames(flow, maxLength, separator)
 		map.addShortFctNames(flow, maxLength, separator)
+		map.addShortMethodNames(flow, maxLength, separator)
```
[the remark this answers is not in the document]
Comment (terfloth) [reply plugins/org.yakindu.sct.model.sexec/src/org/yakindu/sct/model/sexec/naming/DefaultNamingService.xtend]: These are still used. We may remove it once we completely moved from sequences to methods.

**Pull request merged**